### ISSUE TITLE
[question] Random install failures?

### ISSUE THREAD

**Issue by radonish** [conan-io/conan#8241]
In my CI environment I am periodically seeing `conan install` failures as indicated in the log below. For this specific instance it looks like the conanmanifest.txt file failed to properly download. Has anyone seen something similar? Is this a bug or a CI environment configuration issue?

Thanks

**Environment**

- Conan Version: 1.32.1
- Operating System: Red Hat Enterprise Linux 7.4
- Python Version: 3.7.1

**Steps to Reproduce**

1. `conan install`
ie. conan install ../repo/utils/ user/stable -if . -pr=profile-x86_64-rhel

**Log**
```
[2020-12-19T16:30:46.316Z] configutils/1.0.0@user/stable: Not found in local cache, looking in remotes...

[2020-12-19T16:30:46.316Z] configutils/1.0.0@user/stable: Trying with '09570ca2-676b-4b91-bf11-07275d777efa'...

[2020-12-19T16:30:46.329Z] Downloading conanmanifest.txt

[2020-12-19T16:30:46.337Z] Downloading conanfile.py

[2020-12-19T16:30:46.346Z] configutils/1.0.0@user/stable: Downloaded recipe revision c65a9187a96d2179e9a2eafc58d83760

[2020-12-19T16:30:46.590Z] configutils/1.0.0@user/stable: ERROR: Exception while getting package: 5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9

[2020-12-19T16:30:46.590Z] configutils/1.0.0@user/stable: ERROR: Exception: <class 'FileNotFoundError'> [Errno 2] No such file or directory: '/jenkins_server/workspace/build/conanHome/.conan/data/configutils/1.0.0/user/stable/dl/pkg/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/conanmanifest.txt'

[2020-12-19T16:30:46.590Z] ERROR: [Errno 2] No such file or directory: '/jenkins_server/workspace/build/conanHome/.conan/data/configutils/1.0.0/user/stable/dl/pkg/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/conanmanifest.txt'
```

**Comment by radonish** [comment]
Including the backtrace as well:
```
[2020-12-19T16:30:46.520Z] Installing (downloading, building) binaries...

[2020-12-19T16:30:46.520Z] configutils/1.0.0@user/stable: Retrieving package 5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 from remote '09570ca2-676b-4b91-bf11-07275d777efa' 

[2020-12-19T16:30:46.533Z] Downloading conanmanifest.txt

[2020-12-19T16:30:46.548Z] Downloading conaninfo.txt

[2020-12-19T16:30:46.555Z] Downloading conan_package.tgz

[2020-12-19T16:30:46.588Z] Traceback (most recent call last):

[2020-12-19T16:30:46.588Z]   File "/programs/python_3.7.1/lib/python3.7/site-packages/conans/client/command.py", line 2115, in run

[2020-12-19T16:30:46.588Z]     method(args[0][1:])

[2020-12-19T16:30:46.588Z]   File "/programs/python_3.7.1/lib/python3.7/site-packages/conans/client/command.py", line 508, in install

[2020-12-19T16:30:46.588Z]     lockfile_out=args.lockfile_out)

[2020-12-19T16:30:46.588Z]   File "/programs/python_3.7.1/lib/python3.7/site-packages/conans/client/conan_api.py", line 94, in wrapper

[2020-12-19T16:30:46.588Z]     return f(api, *args, **kwargs)

[2020-12-19T16:30:46.588Z]   File "/programs/python_3.7.1/lib/python3.7/site-packages/conans/client/conan_api.py", line 610, in install

[2020-12-19T16:30:46.588Z]     recorder=recorder)

[2020-12-19T16:30:46.588Z]   File "/programs/python_3.7.1/lib/python3.7/site-packages/conans/client/manager.py", line 78, in deps_install

[2020-12-19T16:30:46.588Z]     graph_info=graph_info)

[2020-12-19T16:30:46.588Z]   File "/programs/python_3.7.1/lib/python3.7/site-packages/conans/client/installer.py", line 305, in install

[2020-12-19T16:30:46.588Z]     self._build(nodes_by_level, keep_build, root_node, graph_info, remotes, build_mode, update)

[2020-12-19T16:30:46.588Z]   File "/programs/python_3.7.1/lib/python3.7/site-packages/conans/client/installer.py", line 415, in _build

[2020-12-19T16:30:46.588Z]     self._download(downloads, processed_package_refs)

[2020-12-19T16:30:46.588Z]   File "/programs/python_3.7.1/lib/python3.7/site-packages/conans/client/installer.py", line 399, in _download

[2020-12-19T16:30:46.588Z]     _download(node)

[2020-12-19T16:30:46.588Z]   File "/programs/python_3.7.1/lib/python3.7/site-packages/conans/client/installer.py", line 388, in _download

[2020-12-19T16:30:46.588Z]     self._download_pkg(layout, n)

[2020-12-19T16:30:46.588Z]   File "/programs/python_3.7.1/lib/python3.7/site-packages/conans/client/installer.py", line 403, in _download_pkg

[2020-12-19T16:30:46.588Z]     node.conanfile.output, self._recorder)

[2020-12-19T16:30:46.588Z]   File "/programs/python_3.7.1/lib/python3.7/site-packages/conans/client/remote_manager.py", line 167, in get_package

[2020-12-19T16:30:46.588Z]     self._get_package(layout, pref, remote, output, recorder, info=info)

[2020-12-19T16:30:46.588Z]   File "/programs/python_3.7.1/lib/python3.7/site-packages/conans/client/remote_manager.py", line 187, in _get_package

[2020-12-19T16:30:46.588Z]     package_checksums = calc_files_checksum(zipped_files)

[2020-12-19T16:30:46.588Z]   File "/programs/python_3.7.1/lib/python3.7/site-packages/conans/client/remote_manager.py", line 298, in calc_files_checksum

[2020-12-19T16:30:46.588Z]     for file_name, path in files.items()}

[2020-12-19T16:30:46.588Z]   File "/programs/python_3.7.1/lib/python3.7/site-packages/conans/client/remote_manager.py", line 298, in <dictcomp>

[2020-12-19T16:30:46.588Z]     for file_name, path in files.items()}

[2020-12-19T16:30:46.588Z]   File "/programs/python_3.7.1/lib/python3.7/site-packages/conans/util/files.py", line 145, in md5sum

[2020-12-19T16:30:46.588Z]     return _generic_algorithm_sum(file_path, "md5")

[2020-12-19T16:30:46.588Z]   File "/programs/python_3.7.1/lib/python3.7/site-packages/conans/util/files.py", line 158, in _generic_algorithm_sum

[2020-12-19T16:30:46.588Z]     with open(file_path, 'rb') as fh:

[2020-12-19T16:30:46.588Z] FileNotFoundError: [Errno 2] No such file or directory: '/jenkins_server/workspace/build/conanHome/.conan/data/configutils/1.0.0/user/stable/dl/pkg/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/conanmanifest.txt'

[2020-12-19T16:30:46.588Z] 

Conan build failed
```

**Comment by memsharded** [comment]
Hi @radonish 

Quick question, are you sharing the Conan cache among different concurrent CI processes? From the cache path ``/jenkins_server/workspace/build/conanHome``, it seems you are not generating a clean home for each job. 

The Conan cache is not designed for concurrency, and parallel CI jobs need to use independent caches.

**Comment by radonish** [comment]
Hello @memsharded 

You hit the nail on the head. The cache is indeed being shared between two parallel stages - one builds the new Conan package and another builds/runs the component's unit tests. 

The unit test stage calls CMake directly but is doing a `conan install` to an install-folder. That `conan install` is conflicting with the `conan create` happening in parallel since both commands ultimately trigger the downloading of necessary dependencies.

Would it be safe to do a `conan install` to pull dependencies in, then the parallel `conan create` and `conan install` (using the install-folder parameter) since all of the downloads would have happened in the up-front `conan install` stage?

Thank you

**Comment by beevvy** [comment]
Hi @memsharded, does this comment still apply in current versions?

> The Conan cache is not designed for concurrency, and parallel CI jobs need to use independent caches.

I observed the same intermittent error as OP in a CI environment where the cache is being concurrently shared for package installation (no package removal involved). But according to [these docs](https://docs.conan.io/1/mastering/custom_cache.html#concurrency), my understanding is that it should be safe.

**Comment by memsharded** [comment]
Hi @beevvy 

Yes, Conan 2.0, the cache is still not concurrent. Actually, it is "less" concurrent, because we removed completely some of the partial guards that it had in 1.X. The good news is that the cache in 2.0 is a full redesign, and the way it is intended it will be way more feasible to introduce concurrency in the future. This is a 2.X planned roadmap feature.

Your docs links is pointing to 1.X. In 1.X it might still possible to have some race conditions when installing packages concurrently, for example when installing for different configurations (but some as header-only will be shared among the different configurations). They should more rarely happen, but still possible. One way to try to minimize could be to do a ``conan info`` first to retrieve the common recipes, then doing the parallel ``conan install`` for each configuration later.

**Comment by memsharded** [comment]
Closing in favor of https://github.com/conan-io/conan/issues/15840 where the Conan 2 cache concurrency future progress will be reported.
